doozerlib/cli/get_nightlies.py

```diff
@@ -1,100 +1,212 @@
-        if ignore_arch(arch):
-            continue
-        phase = 'Accepted'
-        if latest:
-            phase = ''
-        nightlies_with_phase[arch] = all_nightlies_in_phase(major, minor, arch, phase)
-
-    primary_coreos_tag = get_primary_container_name(runtime)
-
-    i = 0
-    for x64_nightly, x64_phase in nightlies_with_phase["amd64"]:
-        if i >= limit:
-            break
-        nightly_set = []
-        for arch in util.go_arches:
-            if ignore_arch(arch):
-                continue
-            if latest:
-                nightly, phase = nightlies_with_phase[arch][i]
-            else:
-                nightly, phase = get_closest_nightly(nightlies_with_phase[arch], x64_nightly)
-            nightly_set.append(nightly)
-            nightly_str = f'{nightly} {phase}'
-            if rhcos:
-                if phase != 'Pending':
-                    rhcos = get_coreos_build_from_payload(get_nightly_pullspec(nightly, arch), primary_coreos_tag)
-                    nightly_str += f' {rhcos}'
-            print(nightly_str)
-        print(",".join(nightly_set))
-        i += 1
-
-
-def get_nightly_pullspec(release, arch):
-    suffix = util.go_suffix_for_arch(arch)
-    return f'registry.ci.openshift.org/ocp{suffix}/release{suffix}:{release}'
-
-
-def get_coreos_build_from_payload(payload_pullspec, primary_coreos_tag):
-    """Retrieve the build version of RHCOS (e.g. 411.86.202206131434-0)"""
-    out, err = exectools.cmd_assert(["oc", "adm", "release", "info", "--image-for", primary_coreos_tag, "--", payload_pullspec])
-    if err:
-        raise Exception(f"Error running oc adm: {err}")
-    rhcos_pullspec = out.split('\n')[0]
-    out, err = exectools.cmd_assert(["oc", "image", "info", "-o", "json", rhcos_pullspec])
-    if err:
-        raise Exception(f"Error running oc adm: {err}")
-    image_info = json.loads(out)
-    build_id = image_info["config"]["config"]["Labels"]["version"]
-    return build_id
-
-
-def get_closest_nightly(nightly_with_phase_list, nightly):
-    target_ts = util.get_release_tag_datetime(nightly)
-    min_delta, min_nightly = None, None
-    for nightly, phase in nightly_with_phase_list:
-        nightly_ts = util.get_release_tag_datetime(nightly)
-        delta = target_ts - nightly_ts
-        local_delta = abs(delta.total_seconds())
-        if min_delta is None or local_delta < min_delta:
-            min_delta = local_delta
-            min_nightly = (nightly, phase)
-    return min_nightly
-
-
-def rc_api_url(tag, arch):
+    # find sets of nightlies where all arches have equivalent content
+    eq_sets = []
+    for eq_set in generate_equivalence_sets(nightlies_for_arch):
+        # check for deeper equivalence
+        await eq_set.populate_deeper_equivalence(runtime)
+        if eq_set.deeper_equivalence():
+            eq_sets.append(eq_set)
+            util.green_print(eq_set.details() if details else eq_set)
+            if len(eq_sets) >= limit:
```

Comment:
Ah I see the limit here and the early break. I see the point of generating all sets since if 1 set is not deeply equivalent we won't have any to return.. hmm can we have deeper_equivalence comparison inside generate_equivalence_sets and return early?